TA_dataset/bin/dataset_api.py

```diff
@@ -87,29 +107,53 @@ def ds_lrq_facet_values(
 
 # Executes LRQ run loop of launch-ping-remove API requests until the query completes
 # with a result
+# Returns tuple - value, error message
 def ds_lrq_run_loop(
     client: AuthenticatedClient, body: PostQueriesLaunchQueryRequestBody
 ):
     body.query_priority = PostQueriesLaunchQueryRequestBodyQueryPriority.HIGH
     response = post_queries.sync_detailed(client=client, json_body=body)
+    logger().debug(response)
     result = response.parsed
-    forward_tag = response.headers["x-dataset-query-forward-tag"]
-    steps_done = result.steps_completed
-    steps_total = result.steps_total
-    query_id = result.id
-    while steps_done < steps_total:
-        response = get_queries.sync_detailed(
-            id=query_id,
-            query_type=body.query_type,
-            client=client,
-            last_step_seen=steps_done,
-            forward_tag=forward_tag,
-        )
-        result = response.parsed
+    if result:
+        forward_tag = response.headers["x-dataset-query-forward-tag"]
         steps_done = result.steps_completed
-    delete_queries.sync_detailed(id=query_id, client=client, forward_tag=forward_tag)
+        steps_total = result.steps_total
+        query_id = result.id
+        retry = 0
+        while steps_done < steps_total:
```

Comment:
Some tests for retry functionality (including edge cases would be good), besides that, LGTM

Reply:
There is no complicated logic with exponential backoff or similar. It's very naive implementation. By eyeballing it one more time, I am convinced, that it cannot end up in the infinite loop.